File: lab6/C++/src/Fuel.cpp
#include "Fuel.hpp"
#include "SupplyRequestDetail.hpp"

Fuel::Fuel(const Fuel& other) :
    Resource(rand(), other.GetMaterialType(), other.GetName(), MeasureUnit::eLiter, other.GetQuantity()), m_type(other.m_type) { }

Fuel::Fuel(int id, const std::string& name, FuelType type, float quantity)
    : Resource(id, MaterialType::Fuel, name, MeasureUnit::eLiter, quantity), m_type(type) {
}

Fuel& Fuel::operator = (const Resource& base) {
    if (base.GetMaterialType() != GetMaterialType())
        throw std::invalid_argument("“ип ресурса не соответствует типу экземпл€ра.");
    SetId(base.GetId());
    SetName(base.GetName());
    AssignWarehouse(base.GetWarehouse());
    m_measureUnit = base.GetMeasureUnit();
    m_quantity = base.GetQuantity();
    return *this;
}

bool Fuel::IsMatches(const SupplyRequestDetail& detail) const {
    return detail.GetSupplyType() == MaterialType::Fuel && detail.GetFuelType() == m_type;
}

FuelType Fuel::GetType() const {
    return m_type;
}

File: CourseProject/ILS_WPF/Models/Core/Resources/Fuel.cs
using ILS_WPF.Models.Core.Enums;
using ILS_WPF.Models.Core.Requests;

namespace ILS_WPF.Models.Core.Resources
{
    public class Fuel : Resource
    {
        public FuelType Type { get; set; }

        protected Fuel() { }

        public Fuel(int id, string name, FuelType type, float quantity)
        {
            Id = id;
            Name = name;
            MaterialType = MaterialType.Fuel;
            Type = type;
            Quantity = quantity;
        }

        public override bool IsMatches(SupplyRequestDetail detail)
            => detail.MaterialType == MaterialType.Fuel &&
                   Type == detail.FuelType;
    }
}
